nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/UnpackContent.java

```diff
@@ -154,75 +171,88 @@ protected void init(final ProcessorInitializationContext context) {
         switch (packagingFormat) {
             case TAR_FORMAT:
-                unpacker = new TarUnpacker();
+            case X_TAR_FORMAT:
+                unpacker = tarUnpacker;
                 addFragmentAttrs = true;
                 break;
             case ZIP_FORMAT:
-                unpacker = new ZipUnpacker();
+                unpacker = zipUnpacker;
                 addFragmentAttrs = true;
                 break;
             case FLOWFILE_STREAM_FORMAT_V2:
-                unpacker = new FlowFileStreamUnpacker(new FlowFileUnpackagerV2());
+                unpacker = flowFileStreamV2Unpacker;
                 addFragmentAttrs = false;
                 break;
             case FLOWFILE_STREAM_FORMAT_V3:
-                unpacker = new FlowFileStreamUnpacker(new FlowFileUnpackagerV3());
+                unpacker = flowFileStreamV3Unpacker;
                 addFragmentAttrs = false;
                 break;
             case FLOWFILE_TAR_FORMAT:
-                unpacker = new FlowFileStreamUnpacker(new FlowFileUnpackagerV1());
+                unpacker = flowFileTarUnpacker;
                 addFragmentAttrs = false;
                 break;
-            default:
-                throw new AssertionError("Packaging Format was " + context.getProperty(PACKAGING_FORMAT).getValue());
+            case AUTO_DETECT_FORMAT:
+                // The format of the unpacker should be known before initialization
+                throw new ProcessException(packagingFormat + " is not a valid packaging format");
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+        PackageFormat packagingFormat = PackageFormat.getFormat(context.getProperty(PACKAGING_FORMAT).getValue().toLowerCase());
+        if (packagingFormat == PackageFormat.AUTO_DETECT_FORMAT) {
+            packagingFormat = null;
+            final String mimeType = flowFile.getAttribute(CoreAttributes.MIME_TYPE.key());
+            if (mimeType == null) {
+                logger.error("No mime.type attribute set for {}; routing to failure", new Object[]{flowFile});
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+            }
+
+            for (PackageFormat format: PackageFormat.values()) {
+                if (mimeType.toLowerCase().equals(format.getMimeType())) {
+                    packagingFormat = format;
+                }
+            }
+            if (packagingFormat == null) {
+                logger.info("Cannot unpack {} because its mime.type attribute is set to '{}', which is not a format that can be unpacked; routing to 'success'", new Object[]{flowFile, mimeType});
+                session.transfer(flowFile, REL_SUCCESS);
```

Comment:
Should this be routed to something like a "not unpacked" relationship, rather than success? Seems like outgoing flow files may or may not be unpacked, and you'd have to do some additional logic (like check mime type or something) to see if it worked.